src/util/configInheritance.tsx

```diff
@@ -249,6 +244,31 @@ export const getInheritedRootStorage = (
   return [null, "LXD"];
 };
 
+// export type InheritedDevices =
+//   | InheritedGPU
+//   | InheritedNetwork
+//   | InheritedOtherDevice
+//   | InheritedVolume;
+
+const processInheritedDevices = <T extends LxdDeviceValue>(
+  appliedProfiles: LxdProfile[],
+  inheritedDevices: any[],
+  device: string,
+  filterFn: (device: any) => boolean,
+) => {
+  for (const profile of appliedProfiles) {
+    Object.entries(profile.devices)
+      .filter(([key, device]) => filterFn(device) && key !== "root")
+      .map(([key, dev]) => {
+        inheritedDevices.push({
+          key: key,
+          [device]: dev as T,
+          source: `${profile.name} profile`,
+        });
+      });
+  }
+};
```

Comment:
I think the problem is that in this generic function, we can't ensure the returned values are of a specific device type. I see you tried to solve that with a generic type and passing in a filter function. I think that could work, if the filter function is also using the same type `T`as the return value.

But maybe a simpler approach is to have the generic function collect all devices. In the specific functions, we can focus on filtering and mapping to the correct device type. With this, we don't need to pass the filter around, but apply it where it gets defined.

```suggestion

interface InheritedDevice {
  key: string;
  device: LxdDeviceValue;
  source: string;
}

const getInheritedDevices = (
  values: InstanceAndProfileFormValues,
  profiles: LxdProfile[],
): InheritedDevice[] => {
  const result: InheritedDevice[] = [];
  if (values.entityType === "instance") {
    const appliedProfiles = getAppliedProfiles(values, profiles);
    for (const profile of appliedProfiles) {
      Object.entries(profile.devices).map(([key, device]) => {
        result.push({ key, device, source: `${profile.name} profile` });
      });
    }
  }
  return result;
};

export interface InheritedVolume {
  key: string;
  disk: LxdDiskDevice;
  source: string;
}

export const getInheritedVolumes = (
  values: InstanceAndProfileFormValues,
  profiles: LxdProfile[],
): InheritedVolume[] => {
  return getInheritedDevices(values, profiles)
    .filter((item) => isDiskDevice(item.device) && item.key !== "root")
    .map((item) => ({
      ...item,
      disk: item.device as LxdDiskDevice,
    }));
};
```

Reply:
This is a great solution. I love how you used the spread operator to transform the item object array and also add another property. That's a very cool technique that I will try to keep in mind. Thanks for the implementation :).